File: ui/src/store/datafront/worlds.ts
import { ExplorationData } from '../../domain/misc';
import { type World } from '../../domain/World';
import { WorldClass } from '../../domain/WorldOverview';
import { createDatafrontTable } from '../../lib/datafront/table';
import { type WorldsTableRow } from '../../lib/net/types.generated';
import { ws } from '../../lib/net/ws';
import { updater, cleaner } from './misc';

export const dfWorlds = createDatafrontTable<WorldsTableRow, World>({
    name: 'worlds',
    ws,
    updater,
    cleaner,
    map: (data) => {
        return {
            id: data.worldId,
            params: {
                ageByrs: data.ageByrs ?? 0,
                class: WorldClass.fromString(data.class ?? ''),
                radiusKm: data.radiusKm ?? 0,
                massEarths: data.massEarths ?? 0,
                axisTilt: data.axisTiltRads ?? 0,
                dayLength: data.dayLengthGD ?? 0,
            },
            explored: ExplorationData.parse(data),

            grid: {
                coords: data.gridCoords ?? [],
                edges: data.gridEdges ?? [],
            },
            surface: {
                tempK: data.avgTempK,
                pressureBar: data.surfacePressureBar,
                g: data.g,
            },
            population: {
                bases: data.nBases,
                cities: data.nCities,
                pops: data.nPops,
            },

            elevationsScaleKm: data.elevationsScaleKm,
            elevations: data.elevations ?? [],
            biomes: data.surfaceTypes ?? [],
            colors: (data.colors ?? []).map(([r, g, b]) => ({ r, g, b })),
            oceanLevel: data.oceansLevel,

            atmosphere: data.atmosphere ?? {},
            oceans: data.oceans ?? {},
            snow: data.snow ?? {},
        };
    },
});

File: server/pkg/api/worlds.go
package api

import "time"

const (
	WorldsQueryTypeByID = "byId"
)

type WorldsTableRow struct {
	ID string `json:"worldId"`

	ExploredBy string    `json:"exploredBy,omitempty"`
	ExploredAt time.Time `json:"exploredAt,omitempty"`

	MassEarths float64 `json:"massEarths,omitempty"`
	RadiusKm   float64 `json:"radiusKm,omitempty"`
	AgeByrs    float64 `json:"ageByrs,omitempty"`
	Class      string  `json:"class,omitempty"`

	AxisTiltRads      float64 `json:"axisTiltRads,omitempty"`
	DayLengthGameDays float64 `json:"dayLengthGD,omitempty"`

	GridCoords     []float64   `json:"gridCoords,omitempty"`
	GridEdges      [][]int     `json:"gridEdges,omitempty"`
	Colors         [][]float64 `json:"colors,omitempty"`
	Elevations     []float64   `json:"elevations,omitempty"`
	SurfaceTypes   []string    `json:"surfaceTypes,omitempty"`
	MoistureLevels []float64   `json:"moistureLevels,omitempty"`
	SoilFertility  []float64   `json:"soilFertility,omitempty"`

	ResourceDeposits map[int][]WorldsTableRowResourceDeposit `json:"resources,omitempty"`

	AvgTempK          float64 `json:"avgTempK"`
	PressureBar       float64 `json:"surfacePressureBar"`
	GravityGs         float64 `json:"g"`
	OceansLevel       float64 `json:"oceansLevel"`
	ElevationsScaleKm float64 `json:"elevationsScaleKm"`

	SnowContent       map[string]float64 `json:"snow,omitempty"`
	OceansContent     map[string]float64 `json:"oceans,omitempty"`
	AtmosphereContent map[string]float64 `json:"atmosphere,omitempty"`

	NPops Predictable `json:"nPops"`

	TileCities map[int]int `json:"tileCities"`
	TileBases  map[int]int `json:"tileBases"`
}

type WorldsTableRowResourceDeposit struct {
	ResourceID string  `json:"resource"`
	Abundance  float64 `json:"abundance"`
}
